scipy/stats/unuran/unuran_wrapper.pyi

```diff
@@ -0,0 +1,60 @@
+import numpy as np
+from typing import Union, Any, Protocol, Tuple, List
+import numpy.typing as npt
+
+
+SeedType = Union[np.random.RandomState, np.random.Generator, int,
+                 npt.ArrayLike, np.random.SeedSequence]
+
+
+__all__: List[str]
+
+
+class Method:
+    def rvs(self, size: None | int | tuple = ...) -> np.ndarray: ...
+
+
+class TDRDist(Protocol):
+    def pdf(self, x: float, *args) -> float: ...
+    def dpdf(self, x: float, *args) -> float: ...
+
+
+class TransformedDensityRejection(Method):
+    def __init__(self,
+                 dist: TDRDist,
+                 mode: None | float = ...,
+                 center: None | float = ...,
+                 params: Tuple[Any, ...] = ...,
+                 domain: None | Tuple[float, float] = ...,
+                 c: float = ...,
+                 cpoints: int = ...,
+                 variant: str = ...,
+                 use_dars: bool = ...,
+                 max_sqhratio: float = ...,
+                 max_intervals: int = ...,
+                 use_center: bool = ...,
+                 use_mode: bool = ...,
+                 guide_factor: float = ...,
+                 seed: SeedType = ...) -> None: ...
+
+    @property
+    def sqhratio(self) -> float: ...
+    @property
+    def n_intervals(self) -> int: ...
+    @property
+    def squeeze_area(self) -> float: ...
+
+    def ppf_hat(self, u: npt.ArrayLike) -> float | np.ndarray: ...
```

Comment:
I would strongly advice against returning any unions.
Namely, an operation performed on a union must be compatible with _all_ of its members, 
so users will get into trouble very quickly if they attempt to use it:

``` python
import numpy as np

obj: float | np.ndarray
for i in obj:  # error: Item "float" of "Union[float, ndarray[Any, Any]]" has no attribute "__iter__"
    pass
```

Reply:
I'd instead suggest to use `typing.overload` here to distinguish between 0D and >=1D array-likes:
``` python
from typing import Union, overload
import numpy as np

_ArrayLike0D = Union[bool, int, float, complex, str, bytes, np.generic]

class TransformedDensityRejection(Method):
    @overload
    def ppf_hat(self, u: _ArrayLike0D) -> float: ...
    @overload
    def ppf_hat(self, u: npt.ArrayLike) -> np.ndarray: ...
```

Reply:
Using overload gives me an error:

    Overloaded function signatures 1 and 2 overlap with incompatible return types

This is probably because of different return types. Is there a way to get around this?

Reply:
> This is probably because of different return types. Is there a way to get around this?

I'm afraid this is one of those cases where we'll have to use a `# type: ignore[misc]` comment.

A "proper" solution would be to define a `npt.ArrayLike` sub-type that excludes 0D array-likes,
but without shape-support that's just plain impossible; even with shape-support I fear it will be
tedious enough to the point of just being impractical.

Reply:
Added overloads with `# type: ignore[misc] comments`. Thanks!

Reply:
There are still a bunch of errors in CI that I don't understand:

```
scipy/stats/tests/test_sampling.py:5: error: Module "scipy.stats" has no attribute "TransformedDensityRejection"  [attr-defined]
scipy/stats/tests/test_sampling.py:5: error: Module "scipy.stats" has no attribute "DiscreteAliasUrn"  [attr-defined]
scipy/stats/tests/test_sampling.py:21: error: Module has no attribute "norm"  [attr-defined]
scipy/stats/tests/test_sampling.py:22: error: Module has no attribute "norm"  [attr-defined]
scipy/stats/tests/test_sampling.py:23: error: Module has no attribute "norm"  [attr-defined]
scipy/stats/tests/test_sampling.py:29: error: Module has no attribute "binom"  [attr-defined]
scipy/stats/tests/test_sampling.py:30: error: Module has no attribute "binom"  [attr-defined]
scipy/stats/tests/test_sampling.py:202: error: Module has no attribute "norm"  [attr-defined]
scipy/stats/tests/test_sampling.py:206: error: Module has no attribute "norm"  [attr-defined]
scipy/stats/tests/test_sampling.py:211: error: Module has no attribute "norm"  [attr-defined]
```

If these aren't related, I think we will have to add:

```
[mypy-scipy.stats.tests.test_sampling]
ignore_errors = True
```

in the `mypy.ini` file, right?

Reply:
Yes, though before you change anything: are all these errors related to objects added in this PR?
If so, then it means the annotations are not being exported to `scipy.stats`, which would definitely be worth fixing.

Reply:
> Yes, though before you change anything: are all these errors related to objects added in this PR?

Yes.

> If so, then it means the annotations are not being exported to `scipy.stats`, which would definitely be worth fixing.

I forgot to add `__init__.pyi`. Aster adding that, those errors disappear but some false positives still remain. I added `ignore_errors = True` for them. This should resolve all the mypy errors. Thanks for your help on these!

Reply:
In principle a `__init__.pyi` should be redundant here, as long as the objects are declared 
re-exportable in the `__init__.py` file.

One way to do that is via `import x as x`-style imports or, alternatively, by adding them to a module-level `__all__`.
Outside of stub files I'd say that the latter option is preferred.

Reply:
> In principle a `__init__.pyi` should be redundant here, as long as the objects are declared
> re-exportable in the `__init__.py` file.
> 
> One way to do that is via `import x as x`-style imports or, alternatively, by adding them to a module-level `__all__`.
> Outside of stub files I'd say that the latter option is generally preferred.

Oh, I didn't know that. It worked locally so I have pushed this. Thanks for pointing this out!

Reply:
Thanks for your help on type hints, @BvB93! I have marked most of your comments as resolved. Let me know if you have more comments/reviews.